### PR TITLE
[TOPI][RELAY] Add op Size

### PR DESCRIPTION
**Pull request by yongwww**
Add op Size. Size is used in numpy, TensorFlow and MXNet, is needed for tf object detection models like mask-rcnn.

@zhiics @kevinthesun @icemelon9 @jroesch @Laurawly

### CODE REVIEW

File: src/relay/op/tensor/unary.cc [hunk 1]
```diff
@@ -270,5 +270,54 @@ RELAY_REGISTER_OP("shape_of")
...
+bool SizeRel(const Array<Type>& types,
+             int num_inputs,
+             const Attrs& attrs,
+             const TypeReporter& reporter) {
+    CHECK_EQ(num_inputs, 1);
+    auto tt = types[0].as<TensorTypeNode>();
+    CHECK(tt != nullptr);
+    const auto* param = attrs.as<SizeAttrs>();
+    CHECK(param != nullptr);
+    reporter->Assign(types[1], TensorTypeNode::make({1}, param->dtype));
```
Comment (yzhliu) [inline src/relay/op/tensor/unary.cc]: looks like type of `size` in numpy is scalar?

Comment (yongwww) [reply src/relay/op/tensor/unary.cc]: Yeah, both numpy.size and tf.size are scalar

File: topi/include/topi/transform.h [hunk 2]
```diff
@@ -1156,5 +1156,30 @@ inline Tensor shape(const Tensor& src,
...
+ * \param name output tensor name.
+ * \param tag output tensor tag.
+ * \return Tensor of input shape.
+ */
+inline Tensor size(const Tensor& src,
+                   Type dtype,
+                   const std::string name = "size",
+                   const std::string tag = kInjective) {
+  int ndim = static_cast<int>(src->shape.size());
+  Array<Expr> out_size;
```
Comment (zhiics) [inline topi/include/topi/transform.h]: use `out_size{1}` directly?

File: topi/include/topi/transform.h [hunk 3]
```diff
@@ -1156,5 +1156,30 @@ inline Tensor shape(const Tensor& src,
...
+/*!
+ * \brief Get the size of input tensor.
+ * \param src the input tensor.
+ * \param dtype the type of the elements in the tensor.
+ * \param name output tensor name.
+ * \param tag output tensor tag.
+ * \return Tensor of input shape.
+ */
+inline Tensor size(const Tensor& src,
+                   Type dtype,
```
Comment (zhiics) [inline topi/include/topi/transform.h]: const Type& ?

File: src/relay/op/tensor/unary.cc (hunk 1 above)
Comment (yzhliu) [reply src/relay/op/tensor/unary.cc]: but `{1}` here is tensor? I think we need to keep it the same as np

File: topi/src/topi.cc [hunk 4]
```diff
@@ -300,6 +300,11 @@ TVM_REGISTER_GLOBAL("topi.shape")
   *rv = shape(args[0], args[1]);
 });
 
+TVM_REGISTER_GLOBAL("topi.size")
+.set_body([](TVMArgs args, TVMRetValue *rv) {
+*rv = size(args[0], args[1]);
```
Comment (zhiics) [inline topi/src/topi.cc]: 2 space indent

File: topi/tests/python/test_topi_transform.py [hunk 5]
```diff
@@ -613,6 +613,33 @@ def check_device(device):
...
+    output = np.asarray(np.size(input)).astype(dtype)
+
+    def check_device(device):
+        ctx = tvm.context(device, 0)
+        if not ctx.exist:
+            print("Skip because %s is not enabled" % device)
+            return
+        tvm_input = tvm.nd.array(input, ctx=ctx)
+        tvm_output = tvm.nd.empty((1,), ctx=ctx, dtype=B.dtype)
+        print("Running on target: %s" % device)
```
Comment (zhiics) [inline topi/tests/python/test_topi_transform.py]: remove or log?

File: topi/include/topi/transform.h [hunk 6]
```diff
@@ -1156,5 +1156,30 @@ inline Tensor shape(const Tensor& src,
...
+                   const std::string name = "size",
```
Comment (zhiics) [inline topi/include/topi/transform.h]: either `std::string` or `const std::string&` ?

File: src/relay/op/tensor/unary.cc (hunk 1 above)
Comment (yongwww) [reply src/relay/op/tensor/unary.cc]: good catch, sure I'll update the code

File: tests/python/relay/test_op_level3.py [hunk 7]
```diff
@@ -607,6 +607,24 @@ def verify_gather_nd(xshape, yshape, y_data):
     verify_gather_nd((2, 2, 2), (2, 2), [[0, 1], [1, 0]])
 
 
+def test_size():
```
Comment (Laurawly) [inline tests/python/relay/test_op_level3.py]: Move the test to test_op_level10.py

**Comment by tqchen** [comment]
We need to debate a bit about the API. given that numpy's size is ndarray.size, we may not want to directly use ```relay.size``` as the operator name. If it is a contrib op, consider something like ```relay.contrib.array_size```

**Comment by icemelon9** [comment]
What is the status of this PR? @yongwww @tqchen

**Comment by yongwww** [comment]
@icemelon9 need to address comments, especially the returned scalar representation. I am working on it.

**Comment by jroesch** [comment]
> We need to debate a bit about the API. given that numpy's size is ndarray.size, we may not want to directly use `relay.size` as the operator name. If it is a contrib op, consider something like `relay.contrib.array_size`

I think we should name `tensor_size` given we only talk about Tensors in Relay and not arrays.

**Comment by tqchen** [comment]
@jroesch the naming convention is a bit debatable, given numpy use ndarray and we also use NDArray as our storage type, my guess is that array_size is fine if we want to be relatively consistent with numpy

**Comment by yongwww** [comment]
@tqchen @jroesch both `tensor_size` and `array_size` make sense to me. TensorFlow and NumPy are using `size`. In PyTorch, the corresponding name is `numel`(number of elements of tensor. `size` is used to get the shape of the tensor).  How about  `relay.contrib.numel`? another candidate `relay.contrib.num_element`.

**Comment by tqchen** [comment]
relay.contrib.num_element seems to be a clear name

**Comment by yongwww** [comment]
@yzhliu @Jabher @zhiics @icemelon9 @tqchen have incorporated all comments except for the scalar representation. I tried to add scalar support in my [own branch](https://github.com/yongwww/tvm/tree/size).  But failed to run on graph runtime since the schedule was mutated in Halide improperly.  Filed [an issue](https://discuss.tvm.ai/t/return-scalar-value-for-some-ops/2939).

**Comment by jroesch** [comment]
Can you rebase?

File: topi/tests/python/test_topi_transform.py (hunk 5 above)
Comment (yongwww) [reply topi/tests/python/test_topi_transform.py]: simply use print here to make it consistent with all the other tests in topi. All test cases are using print.

**Comment by yongwww** [comment]
@jroesch @zhiics rebased and fixed zhi's comments.

**Comment by a1010428282** [comment]
I wonder if this commit will continue？

**Comment by tqchen** [comment]
@yongwww please rebase, rename to ndarray_size and let us merge this

**Comment by yongwww** [comment]
@tqchen renamed. Pls take a look

**Comment by tqchen** [comment]
Please rename the topi api name as well.

**Comment by yongwww** [comment]
@tqchen updated. Seems CI hangs
